skt/runner.py

```diff
@@ -281,16 +281,29 @@ def get_mfhost(self):
         return max(set(fhosts), key=fhosts.count)
 
     def __jobresult(self, jobid):
+        """
+        Get results for a specific job.
+
+        Args:
+            jobid: A Beaker job ID, in 'J:<id>' format.
+
+        Returns:
+            A tuple (ret, result), where ret is 0 if the job ended with Pass,
+            1 if a test failure occurred and 2 if it aborted (infrastructure
+            failure); and result is the job result itself.
+        """
         ret = 0
         result = None
 
-        if jobid is not None:
-            root = self.getresultstree(jobid)
-            result = root.attrib.get("result")
+        root = self.getresultstree(jobid)
+        result = root.attrib.get("result")
+        status = root.attrib.get('status')
 
-            if result != "Pass":
-                ret = 1
-            logging.info("job result: %s [%d]", result, ret)
+        if result == 'Warn' and status == 'Aborted':
+            ret = 2
+        elif result != "Pass":
+            ret = 1
```

Comment:
Can we instead check for known "Pass" and "Fail" conditions explicitly, and consider everything else an error and return 2? This way we'll keep all the weird results away from developers, and get to learn all possible failure modes. I.e. if Beaker produces anything unanticipated we get to investigate and decide whether that is a "Pass" or a "Fail", instead of just assuming it's a "Fail" and sending the message?

Reply:
Not really. 'Panic' is a separate result (same as 'Pass' and 'Fail'), so doing so we wouldn't be reporting kernel panics. Also, 'Warn' itself is a valid test result. Only the combination of 'Warn' and 'Aborted' should show infrastructure failures.

I totally understand if this is confusing. I wasn't able to find any useful Beaker documentation and had to dig through a bunch of stuff to find out what's going on.

Reply:
> I totally understand if this is confusing. I wasn't able to find any useful Beaker documentation and had to dig through a bunch of stuff to find out what's going on.

Perhaps we should write our own documentation, then. Right here, in the code, maybe?

Otherwise, how about we rewrite it this way, then:
```
if result == "Pass":
    ret = 0
elif result != "Warn" or status != "Aborted":
    ret = 1
else:
    ret = 2
```
and perhaps narrow the "Fail" (`ret = 1`) condition down a bit, if we can?
We won't have to initialize `ret` too.

Reply:
The warn/aborted condition has to be `and`, not `or`, but otherwise I'm fine with rewriting it to this style.

Reply:
Hmm, I just reverted the condition using the `!(A&&B) => !A||!B` rule. Did I miss something?

Reply:
Oh, right. I got confused by the comparison of my version and this one, and didn't notice that the switch is setting 1 and not 2. Since it kind of makes more sense to me to have the specific condition in the `if ` and all the rest in the `else` (so `if specific infrastructure failure else 1`, if that makes sense).

It should work then, thanks for checking me :)

Reply:
OK, great :) Now, can we narrow that "ret=1" condition further somehow?

Reply:
I don't think so. For the completed jobs, you can get 'Pass', 'Warn', 'Fail' and 'Panic'. 'Panic' is for kernel panics, which we should report, 'Fail' and 'Warn' are normal test results which mean something went wrong.

Reply:
Ah, OK, then. Do you know what's the difference between "Warn" and "Fail"? And why do we even have "status" in addition to "result"?

Reply:
What they mean is test-specific. The test author decides what s/he considers a warning and failure. For example, I've seen NFS tests which considered a warning if the tests completed, but the speed was under a specific threshold; and a failure if they didn't complete.

Well, 'result' is the result of the job and 'status' it's status. Results are those 'Pass', 'Fail', 'Warn' things, while status talks about whether the job completed or not - status can be eg. 'Completed' and 'Aborted', 'Scheduled' (most of them aren't relevant once the job is finished).

Reply:
I feel there are some very dark corners there, so let's leave the condition at that for now. Thank you for all the explanations :)